### ISSUE TITLE
Removal of outdated packages with out paccache

### ISSUE THREAD

**Issue by chrishoage** [nroi/flexo#39]
Thank you for this wonderful tool! 

I'm using https://github.com/docker-flexo/docker-flexo which is based on debian, and I'm hosting it on an ubuntu sever. 

Furthermore I'm using a named docker-compose volume to store the cache like so 

```yml
  flexo:
    restart: ${RESTART-unless-stopped}
    image: nroi/flexo
    container_name: flexo
    volumes:
      - flexo-cache:/var/cache/flexo
```

All this put together makes it a challenge to run `paccache -r`. Currently the best solution is to spin up a temporary docker container using the archlinux image, mount the named volume, and run the command in the README using `paccache -r`. 

While this works it would be nice if flexo would prune old packages on its own with out the use of `paccache -r`. 

Is this in scope?

**Comment by nroi** [comment]
Thank you for your feedback!

I understand that using paccache is not ideal when Flexo runs inside docker, but I would like to avoid duplicating the functionality of paccache inside Flexo.

One way to allow Flexo to purge outdated packages would be the following:
* Flexo provides two new HTTP endpoints: one `GET` endpoint to fetch a list of package names that are currently cached by Flexo, and another `POST` endpoint where Flexo receives a list of file names.
* An ArchLinux client can then fetch the list of all cached packages, run `paccache` against this list to obtain all files that can be purged, and then `POST` the result to Flexo.
* Flexo receives this result and removes all files included in this POST request (of course, for security reasons, Flexo will make sure that no files outside the cache directory can be deleted, so you won't be table to use that `POST` request to delete sensitive files).

With this approach, the entire logic on how and when to purge packages resides on the client-side. I would then provide a new package on AUR that includes a small script (probably python) that does all this (i.e., running paccache and sending the HTTP GET and POST requests).

The downside is that you would need to install an additional package on one of your clients, and you would now have to maintain two configuration files for Flexo: One configuration file for the Flexo server (the existing TOML file), and an additional config file on a client. The client-side config file includes the host name of the Flexo server.

Is this something you would use, or would you consider this even more inconvenient than your current "archlinux inside docker" approach?

If you (or anyone else) has any other ideas or thoughts on this, feel free to add them.

**Comment by chrishoage** [comment]
> Is this something you would use, or would you consider this even more inconvenient than your current "archlinux inside docker" approach?

To be entirely honest, no I wouldn't use that and would continue to just spin up a temporary docker container.

I would ask you to consider the viewpoint of a consumer of this tool: By default this tool will never purge cached packages with out manual intervention. For a tool that manages a cache, a consumer might expect purging expired caches to be built in. Imagine a CDN with no expiration policy requiring manual intervention.  

> With this approach, the entire logic on how and when to purge packages resides on the client-side. I would then provide a new package on AUR that includes a small script (probably python) that does all this (i.e., running paccache and sending the HTTP GET and POST requests).

I absolutely understand the desire to not duplicate functionality of an existing tool, and I obviously don't understand the architecture of this tool you have created. However your suggestion seems a _great deal_ more complex than to [implement pacsort](https://git.archlinux.org/pacman-contrib.git/tree/src/pacsort.c) and delete the results older than a configured set (with a sane value set by default). The [actual paccache utility](https://git.archlinux.org/pacman-contrib.git/tree/src/paccache.sh.in) is quite simple (compared to pacsort) and _feels_ simpler than extra HTTP endpoints, and exposing unauthenticated REST endpoints that delete files (regardless of any security put in place to limit where it can delete)

> If you (or anyone else) has any other ideas or thoughts on this, feel free to add them.

Since this is largely a problem for the docker container, and largely wouldn't be an issue for most users running it natively (it is unlikely people would be running it on metal on anything other than Arch), one other solution would be to instead base the docker image on Arch instead of debain and include pacman-contrib in the image. 

This way, users of the docker container could shell into the flexo container, run a paccache command, and exit the shell. 

This sidesteps the issue, but would be simpler than both other suggestions.

**Comment by nroi** [comment]
I decided to base the image on debian because those images are relatively small (compared to ArchLinux) and because I prefer to use a non-rolling-release distro for Docker images. With ArchLinux, errors will be much more difficult to reproduce because two subsequent docker builds will not necessarily contain the same application software versions, even if the Dockerfile hasn't changed. For example, the current docker image installs curl on Debian. With ArchLinux, it's impossible to control what version of curl is installed.

So this leaves us with the question of how to deploy *some* implementation of `paccache` on Debian. The `paccache` bash script itself is relatively simple, I admit, but it depends on pacman-contrib which in turn depends on libalpm from pacman. For example, [version.c](https://git.archlinux.org/pacman.git/tree/lib/libalpm/version.c) would also have to be reimplemented.

I'm going to look into this more closely when I find the time, but don't expect this to be solved within just a few days… If anyone wants to support me with this, this is how I would start:

Create a paccache clone in some higher-level language of your choice (**NOT** C or C++). This paccache-clone does not require the entire functionality of paccache, it only needs to be able to scan the filesystem and, with a parameter `k`, print all files that can be deleted in order to keep only the `k` most recent versions. This paccache-clone should build and run without any elaborate setup on the `debian:buster-slim` docker image, so it must not require pacman, or libalpm, or anything else that cannot be installed without difficulties on Debian.

When this paccache-clone exists, then I can modify Flexo to run this program after an HTTP GET request was served, so that after each download, all obsolete packages in the cache are purged.

**Comment by chrishoage** [comment]
In the mean time, I've created an Archlinux based image that uses cronie and s6 to clean the cache weekly (the corn tab is configurable). It is a great deal larger than the image provided at `nori/docker-flexo` (~550mb compared to ~50mb) so if that is a concern please do consider it. 

https://github.com/chrishoage/docker-arch-flexo

**Comment by nroi** [comment]
I've just released a first version of [Scruffy](https://github.com/nroi/scruffy), it's basically an alternative to `paccache` that does not require pacman/libalpm. It just has the relevant bits of C-code copied from pacman. I've decided that this is more sensible than a complete reimplementation of `alpm_pkg_vercmp` because the code includes a lot of pointer arithmetic which cannot be translated into idiomatic Rust.

I'm going to see how to integrate Scruffy into Flexo's debian image so that Docker users can have the cache cleaned regularly.

**Comment by nroi** [comment]
This feature is now available with Flexo version 1.2.2.

If you use Docker, make sure that the Docker image is tagged with a version of 1.2.2 or higher. The default behavior is to keep 3 versions. If you want to change this, set the `FLEXO_NUM_VERSIONS_RETAIN` environment variable.

If you have installed Flexo locally, adapt your `/etc/flexo/flexo.toml` file by including the `num_versions_retain` setting. Check the [example config](https://github.com/nroi/flexo/blob/master/flexo/conf/flexo.toml) for more details.
